pyfar/dsp/regularization.py

```diff
@@ -0,0 +1,408 @@
+"""
+Documentation for regularization class.
+"""
+
+import pyfar as pf
+from pyfar.dsp.dsp import _cross_fade
+import numpy as np
+import numbers
+
+class RegularizedSpectrumInversion():
+    r"""Class for frequency-dependent regularized inversion.
+
+    Regularized inversion limits the gain applied by an inverse filter.
+    This can be useful to avoid extreme amplification when inverting a signal
+    of limited bandwidth or containing notches in the spectrum.
+
+    The inverse is computed as [#]_:
+
+    .. math::
+        :label: regularized_inversion
+
+        S^{-1}(f) = \frac{S^*(f)}{S^*(f)S(f) + \beta |\epsilon(f)|^2} D(f)
+
+    with :math:`f` being the frequency, :math:`S(f)` the spectrum of the signal
+    to be inverted, :math:`(\cdot)^*` the complex conjugate,
+    :math:`\epsilon(f)` the regularization, and :math:`\beta` a scalar to
+    control the amount of regularization. :math:`D(f)` denotes an optional
+    target function.
+
+    The compensated system :math:`C = S(f)S^{-1}(f)` approaches the target
+    function in magnitude and phase. In many applications, the target function
+    should contain a delay to make sure that :math:`S^{-1}(f)` is causal. The
+    larger :math:`\beta` and :math:`\epsilon(f)` are, the larger the deviation
+    of :math:`C(f)` from the target :math:`D(f)`.
+
+    The inversion is done in two steps:
+
+    1. Define :math:`S(f)`, :math:`\epsilon(f)`, :math:`\beta`, and
+       :math:`D(f)` using one of the ``from_...()`` methods listed below.
+    2. Compute the inverse :math:`S(f)^{-1}` using :py:func:`~invert`
+
+    The parameters that are defined in the first step are often iteratively
+    adjusted. Examples are given in the documentation of the ``from_...()``
+    methods.
+
+    References
+    ----------
+    .. [#]  S. G. Norcross, M. Bouchard, and G. A. Soulodre, “Inverse Filtering
+            Design Using a Minimal-Phase Target Function from Regularization,”
+            Convention Paper 6929 Presented at the 121st Convention, 2006
+            October 5–8, San Francisco, CA, USA.
+    """
+
+    def __init__(self) -> None:
+
+        # throw error if object is instanced without classmethod
+        raise RuntimeError("Regularization objects must be created "
+                           "using one of the 'from_()' classmethods.")
+
+    def __repr__(self):
+        """String representation of RegularizedSpectrumInversion class."""
+        return f"Regularization object with regularization "\
+            f"of type '{self._regularization_type}'."
+
+    @classmethod
+    def from_frequency_range(cls, signal, frequency_range,
+                             regularization_within=0, beta=1, target=None):
+        r"""
+        Regularization from a given frequency range.
+
+        Defines a frequency range within which the regularization
+        :math:`\epsilon(f)` is set to `regularization_within`.
```

Comment:
```suggestion
        :math:`\epsilon(f)` is set to ``regularization_within``.
```

our guidelines are suggesting double ``